modules/apps/29-fee/keeper/escrow.go

```diff
@@ -85,31 +86,36 @@ func (k Keeper) DistributePacketFeesOnTimeout(ctx sdk.Context, timeoutRelayer sd
 		}
 
 		// refund receive fee for unused forward relaying
-		k.distributeFee(ctx, refundAddr, feeInEscrow.Fee.RecvFee)
+		k.distributeFee(ctx, refundAddr, refundAddr, feeInEscrow.Fee.RecvFee)
 
 		// refund ack fee for unused reverse relaying
-		k.distributeFee(ctx, refundAddr, feeInEscrow.Fee.AckFee)
+		k.distributeFee(ctx, refundAddr, refundAddr, feeInEscrow.Fee.AckFee)
 
 		// distribute fee for timeout relaying
-		k.distributeFee(ctx, timeoutRelayer, feeInEscrow.Fee.TimeoutFee)
+		k.distributeFee(ctx, timeoutRelayer, refundAddr, feeInEscrow.Fee.TimeoutFee)
 	}
 }
 
 // distributeFee will attempt to distribute the escrowed fee to the receiver address.
 // If the distribution fails for any reason (such as the receiving address being blocked),
 // the state changes will be discarded.
-func (k Keeper) distributeFee(ctx sdk.Context, receiver sdk.AccAddress, fee sdk.Coins) {
+func (k Keeper) distributeFee(ctx sdk.Context, receiver, refundAccAddress sdk.AccAddress, fee sdk.Coins) {
 	// cache context before trying to distribute fees
 	cacheCtx, writeFn := ctx.CacheContext()
 
 	err := k.bankKeeper.SendCoinsFromModuleToAccount(cacheCtx, types.ModuleName, receiver, fee)
-	if err == nil {
-		// write the cache
-		writeFn()
-
-		// NOTE: The context returned by CacheContext() refers to a new EventManager, so it needs to explicitly set events to the original context.
-		ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
+	if err != nil && !bytes.Equal(receiver, refundAccAddress) {
```

Comment:
I think you need to separate these conditions as well. What happens if there is an err and refundAcc == receiver?

Maybe something like:
```go
if err != nil {
    // fee distribution failed, try to refund to original incentivizer
    if bytes.Equal(receiver, refundAccAddress) {
        return err
    }
    
    // attempt to distribute fee to refund address
}
```

Reply:
Could we add a test case for this case?

Reply:
If there's an error and `refundAcc == receiver` then sending to the refundAcc has already failed, right? 
That's why I combined the conditions using the logical AND.
The issues writes:

> If we are trying to refund to original sender and that fails, then we need to just no-op and continue.

In this case, `writeFn()` is still called but there would be no state changes to commit (no-op).
I can adapt this tomorrow tho if we feel its better separated.

I'll look at adding a testcase 👍

Reply:
Tests require some refactoring. In progress!

Reply:
To be on the safe side I think it makes sense to ensure `writeFn()` doesn't get called. Otherwise we need to verify that no partial state transitions occurred in x/bank. It might assume state changes are discarded once the error is returned

Reply:
Sure, I'll adapt the logic to represent that. I had similar concerns wrt `writeFn()` but wasn't 100%. Thanks :)